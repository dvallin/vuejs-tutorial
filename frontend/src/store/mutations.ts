import {Task, TaskState} from "../domain/Task";

export enum Mutations {
  AddTaskSucceeded = 'addTaskSucceeded',
  TaskStateChanged = 'taskStateChanged',
  FetchTasksSucceeded = 'fetchTasksSucceeded',

  FetchTasksFailed = 'fetchTasksFailed',
  TaskNotFound = 'taskNotFound',
  AddTaskFailed = 'addTaskFailed'
}

export interface AddTaskSucceeded {
  type: Mutations.AddTaskSucceeded;
  task: Task;
}
export interface TaskStateChanged {
  type: Mutations.TaskStateChanged;
  task: Task;
  newTaskState: TaskState;
}
export interface FetchTasksSucceeded {
  type: Mutations.FetchTasksSucceeded;
  tasks: Array<Task>;
}
export interface FetchTasksFailed {
  type: Mutations.FetchTasksFailed;
  error: string;
}
export interface TaskNotFound {
  type: Mutations.TaskNotFound;
  taskTitle: string;
}
export interface AddTaskFailed {
  type: Mutations.AddTaskFailed;
  taskTitle: string;
}

export const mutations = {
  [Mutations.AddTaskSucceeded] (state, payload: AddTaskSucceeded) {
    state.tasks.push(payload.task);
  },
  [Mutations.TaskStateChanged] (state, payload: TaskStateChanged) {
    payload.task.state = payload.newTaskState;
  },
  [Mutations.FetchTasksSucceeded] (state, payload: FetchTasksSucceeded) {
    state.tasks = payload.tasks;
  },
  [Mutations.TaskNotFound] (state, payload: TaskStateChanged) {
    state.errors.push(payload);
  },
  [Mutations.FetchTasksFailed] (state, payload: FetchTasksFailed) {
    state.errors.push(payload);
  },
  [Mutations.AddTaskFailed] (state, payload: AddTaskFailed) {
    state.errors.push(payload);
  }
};
