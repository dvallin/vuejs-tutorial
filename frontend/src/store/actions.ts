import _ from 'lodash';

import {Task, TaskState} from "../domain/Task";
import {Mutations} from "./mutations";

export enum Actions {
  AddTask = 'addTask',
  SetTaskComplete = 'setTaskComplete'
}

export interface AddTask {
  type: Actions.AddTask;
  title: string;
}

export interface SetTaskComplete {
  type: Actions.SetTaskComplete;
  title: string;
  complete: boolean;
}

export const actions = {
  [Actions.AddTask] ({getters, commit}, payload: AddTask) {
    if(_.isNil(getters.taskByTitle(payload.title))) {
      commit({
        type: Mutations.TaskAdded,
        task: new Task(payload.title)
      });
    }
  },
  [Actions.SetTaskComplete] ({getters, commit}, payload: SetTaskComplete) {
    const task = getters.taskByTitle(payload.title);
    if(_.isNil(task)) {
      commit({
        type: Mutations.TaskNotFound,
        title: payload.title
      });
    } else {
      const newTaskState = payload.complete ? TaskState.Done : TaskState.Open;
      commit({
        type: Mutations.TaskStateChanged,
        task,
        newTaskState
      })
    }
  }
};
