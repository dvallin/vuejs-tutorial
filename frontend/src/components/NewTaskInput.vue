<template>
  <input class="new-todo"
         autofocus autocomplete="off"
         placeholder="What needs to be done?"
         v-model="taskTitle"
         @keyup.enter="addTask">
</template>

<script lang="ts">
  import {Actions} from "../store/actions";

  export default {
    name: 'NewTaskInput',
    data() {
      return { taskTitle: "" }
    },
    methods: {
      addTask() {
        const value = this.taskTitle && this.taskTitle.trim();
        this.$store.dispatch({
          type: Actions.AddTask,
          title: value
        })
      }
    }
  }
</script>
