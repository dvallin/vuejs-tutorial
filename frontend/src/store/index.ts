import * as Vuex from 'vuex';
import Vue from 'vue';

import {mutations} from './mutations';
import {actions} from "./actions";

import {Task} from "@/domain/Task";

export const initialState = () => {
  return {
    tasks: [new Task("new task"), new Task("old task")],
    errors: []
  }
};

export const getters = {
  taskByTitle: (state, getters) => (title) => {
    return state.tasks.find(task => task.title === title)
  }
};

Vue.use(Vuex);
export default new Vuex.Store({
  state: initialState(),
  getters,
  mutations,
  actions
});
