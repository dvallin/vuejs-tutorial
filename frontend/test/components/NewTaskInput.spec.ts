import NewTaskInput from '../../src/components/NewTaskInput'
import {Actions} from "../../src/store/actions";
import store from '../../src/store'
import {mount, shallow} from 'vue-test-utils'

describe('NewTaskInput.vue', () => {
  it('renders an input field', () => {
    const wrapper = shallow(NewTaskInput);
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.find('input')).toBeDefined();
  });

  describe('add task', () => {
    it('adds task from trimmed value of input field, if enter key is pressed', (done) => {
      const spy = jest.spyOn(store, 'dispatch');
      const givenValue = "A very special value    ";

      const wrapper = mount(NewTaskInput, { store });
      const input = wrapper.find('input');
      input.element.value = givenValue;
      input.trigger('input');
      input.trigger('keyup.enter');

      wrapper.vm.$nextTick(() => {
        expect(spy).toHaveBeenCalledWith({
          type: Actions.AddTask,
          title: givenValue.trim()
        });
        done();
      });
    });
  });
});
