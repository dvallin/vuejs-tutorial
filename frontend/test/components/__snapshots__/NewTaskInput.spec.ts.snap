// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTaskInput.vue renders an input field 1`] = `
<input
  autocomplete="off"
  autofocus="autofocus"
  class="new-todo"
  placeholder="What needs to be done?"
/>
`;
